File: android/app/src/main/java/com/lightningkite/kwifttemplate/shared/views/SliderDemoVG.kt
package com.lightningkite.kwifttemplate.shared.views

import android.view.View
import com.lightningkite.kwift.observables.actual.bind
import com.lightningkite.kwift.observables.actual.bindFloat
import com.lightningkite.kwift.observables.actual.bindText
import com.lightningkite.kwift.observables.shared.MutableObservableProperty
import com.lightningkite.kwift.observables.shared.StandardObservableProperty
import com.lightningkite.kwift.observables.shared.transformed
import com.lightningkite.kwift.views.actual.ViewDependency
import com.lightningkite.kwift.views.shared.ViewGenerator
import com.lightningkite.kwifttemplate.xml.SliderDemoXml

class SliderDemoVG() : ViewGenerator() {
    override val title: String get() = "Slider Demo"

    val ratio: StandardObservableProperty<Float> = StandardObservableProperty(0f)
    val percent: MutableObservableProperty<Int> = ratio.transformed(
        read = { it -> (it * 100).toInt() },
        write = { it -> it.toFloat() / 100f }
    )
    val obsRatingInt: MutableObservableProperty<Int> = ratio.transformed(
        read = { it -> (it * 5).toInt() },
        write = { it -> it.toFloat() / 5f }
    )
    val obsRatingFloat: MutableObservableProperty<Float> = ratio.transformed(
        read = { it -> it * 5f },
        write = { it -> it / 5f }
    )

    override fun generate(dependency: ViewDependency): View {
        val xml = SliderDemoXml()
        val view = xml.setup(dependency)

        xml.slider.bind(0, 100, percent)
        xml.valueDisplay.bindText(percent) { it -> it.toString() }

        xml.rating.bind(5, obsRatingInt)
        xml.ratingDisplayStars.bind(5, obsRatingInt)
        xml.ratingDisplayStarsSmall.bind(5, obsRatingInt)
        xml.ratingDisplayNumber.bindText(obsRatingInt) { it -> it.toString() }

        xml.ratingFloat.bindFloat(5, obsRatingFloat)
        xml.ratingDisplayStarsFloat.bindFloat(5, obsRatingFloat)
        xml.ratingDisplayStarsSmallFloat.bindFloat(5, obsRatingFloat)
        xml.ratingDisplayNumberFloat.bindText(obsRatingFloat) { it -> it.toString() }

        return view
    }
}

File: iosOld/com/lightningkite/kwifttemplate/shared/views/SliderDemoVG.swift
//Package: com.lightningkite.kwifttemplate.shared.views
//Converted using Kwift2

import Foundation



public class SliderDemoVG: ViewGenerator {
    
    
    override public var title: String {
        get {
            return "Slider Demo"
        }
    }
    public var ratio: StandardObservableProperty<Float>
    public var percent: MutableObservableProperty<Int32>
    public var obsRatingInt: MutableObservableProperty<Int32>
    public var obsRatingFloat: MutableObservableProperty<Float>
    
    override public func generate(dependency: ViewDependency) -> View {
        var xml = SliderDemoXml()
        var view = xml.setup(dependency)
        xml.slider.bind(0, 100, percent)
        xml.valueDisplay.bindText(percent) { (it) in 
            it.toString()
        }
        xml.rating.bind(5, obsRatingInt)
        xml.ratingDisplayStars.bind(5, obsRatingInt)
        xml.ratingDisplayStarsSmall.bind(5, obsRatingInt)
        xml.ratingDisplayNumber.bindText(obsRatingInt) { (it) in 
            it.toString()
        }
        xml.ratingFloat.bindFloat(5, obsRatingFloat)
        xml.ratingDisplayStarsFloat.bindFloat(5, obsRatingFloat)
        xml.ratingDisplayStarsSmallFloat.bindFloat(5, obsRatingFloat)
        xml.ratingDisplayNumberFloat.bindText(obsRatingFloat) { (it) in 
            it.toString()
        }
        return view
    }
    override public func generate(_ dependency: ViewDependency) -> View {
        return generate(dependency: dependency)
    }
    
    override public init() {
        self.ratio = StandardObservableProperty(0)
        self.percent = ratio.transformed(read: { (it) in 
            ( it * 100 ).toInt()
        }, write: { (it) in 
            it.toFloat() / 100
        })
        self.obsRatingInt = ratio.transformed(read: { (it) in 
            ( it * 5 ).toInt()
        }, write: { (it) in 
            it.toFloat() / 5
        })
        self.obsRatingFloat = ratio.transformed(read: { (it) in 
            it * 5
        }, write: { (it) in 
            it / 5
        })
        super.init()
    }
}
 
